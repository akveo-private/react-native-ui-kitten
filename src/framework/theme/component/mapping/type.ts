// TODO(mapping/type): declare Config/Token type

export type ThemeMappingConfigType = any;

export interface ThemeMappingType {
  parameters: string[];
  variants: any;
}

export interface VariantType {
  state: any;
}

export type TokenType = any;
