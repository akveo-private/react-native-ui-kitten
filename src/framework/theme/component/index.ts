export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
