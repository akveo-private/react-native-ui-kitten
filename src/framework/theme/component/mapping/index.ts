export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
