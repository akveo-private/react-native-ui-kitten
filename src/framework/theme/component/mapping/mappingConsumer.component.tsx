import React, { ComponentType } from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';
import {
  getComponentThemeMapping,
  forwardProps,
} from '../../service';

export interface Props<P> extends React.ClassAttributes<P> {
  mapping: ThemeMappingType;
}

export function withMapping<P extends Props<P>>(Component: ComponentType<P>) {
  type TExcept = Exclude<keyof P, keyof Props<P>>;
  type ForwardedProps = Pick<P, TExcept>;

  class Wrapper extends React.Component<ForwardedProps> {
    wrappedComponentRef = undefined;
    getWrappedInstance = undefined;

    setWrappedComponentRef = (ref) => {
      this.wrappedComponentRef = ref;
    };

    getComponentName = () => Component.displayName || Component.name;

    renderWrappedComponent = (mapping: ThemeMappingType[]) => (
      <Component
        ref={this.setWrappedComponentRef}
        mapping={getComponentThemeMapping(this.getComponentName(), mapping)}
        {...this.props}
      />
    );

    render() {
      return (
        <MappingContext.Consumer>
          {this.renderWrappedComponent}
        </MappingContext.Consumer>
      );
    }
  }

  const Result = Wrapper;
  Result.prototype.getWrappedInstance = function getWrappedInstance() {
    const hasWrappedInstance = this.wrappedComponentRef && this.wrappedComponentRef.getWrappedInstance;
    return hasWrappedInstance ? this.wrappedComponentRef.getWrappedInstance() : this.wrappedComponentRef;
  };

  return forwardProps(Component, Result);
}
