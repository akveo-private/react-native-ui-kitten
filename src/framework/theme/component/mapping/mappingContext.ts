import React from 'react';
import { ThemeMappingType } from './type';

const defaultValue: ThemeMappingType = {};
const MappingContext = React.createContext(defaultValue);

export default MappingContext;
