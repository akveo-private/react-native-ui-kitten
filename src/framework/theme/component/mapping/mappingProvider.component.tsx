import React from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';

export interface Props {
  mapping: ThemeMappingType;
  children: JSX.Element | React.ReactNode;
}

export class MappingProvider extends React.PureComponent<Props> {

  render() {
    return (
      <MappingContext.Provider
        value={this.props.mapping}>
        {this.props.children}
      </MappingContext.Provider>
    );
  }
}
