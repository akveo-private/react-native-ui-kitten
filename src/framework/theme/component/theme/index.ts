export * from './type';
export {
  ThemeProvider,
  Props as ThemeProviderProps,
} from './themeProvider.component';
export {
  withStyles,
  Props as ThemedComponentProps,
} from './themeConsumer.component';
