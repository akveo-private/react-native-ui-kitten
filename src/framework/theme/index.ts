export {
  StyleProvider,
  styled,
  ThemeProvider,
  withStyles,

  StyleProviderProps,
  StyledComponentProps,
  ThemeProviderProps,
  ThemedComponentProps,

  ThemeType,
  ThemeMappingType,
  ComponentMappingType,
  TokenType,
  ThemedStyleType,
  StyleSheetType,
  StyleType,
} from './component';

export {
  APPEARANCE_DEFAULT,
} from './service';

