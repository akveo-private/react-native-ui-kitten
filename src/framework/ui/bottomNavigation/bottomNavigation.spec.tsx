import React from 'react';
import {
  Image,
  ImageProps,
  ImageSourcePropType,
} from 'react-native';
import {
  fireEvent,
  render,
  RenderAPI,
  shallow,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import {
  styled,
  ApplicationProvider,
  ApplicationProviderProps,
  StyleType,
} from '@kitten/theme';
import {
  BottomNavigation as BottomTabNavigatorComponent,
  Props as BottomTabNavigatorProps,
} from './bottomNavigation.component';
import {
  BottomNavigationTab as BottomNavigatorTabComponent,
  Props as BottomNavigatorTabProps,
} from './bottomNavigationTab.component';
import { default as mapping } from '../common/mapping.json';
import { default as theme } from '../common/theme.json';

const BottomNavigation = styled<BottomTabNavigatorProps>(BottomTabNavigatorComponent);
const BottomNavigationTab = styled<BottomNavigatorTabProps>(BottomNavigatorTabComponent);

describe('@bottom-navigation-tab: component checks', () => {

  const iconSource: ImageSourcePropType = { uri: 'https://akveo.github.io/eva-icons/fill/png/128/star.png' };

  const Mock = (props?: BottomNavigatorTabProps): React.ReactElement<ApplicationProviderProps> => {
    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <BottomNavigationTab {...props}/>
      </ApplicationProvider>
    );
  };

  const icon = (style: StyleType): React.ReactElement<ImageProps> => {
    return (
      <Image
        style={style}
        source={iconSource}
      />
    );
  };

  it('* empty', () => {
    const component: RenderAPI = render(
      <Mock/>,
    );

    const { output } = shallow(component.getByType(BottomNavigatorTabComponent));

    expect(output).toMatchSnapshot();
  });

  it('* with icon', () => {
    const component: RenderAPI = render(
      <Mock icon={icon}/>,
    );

    const { output } = shallow(component.getByType(BottomNavigatorTabComponent));

    expect(output).toMatchSnapshot();
  });

  it('* text/selected', () => {
    const component: RenderAPI = render(
      <Mock
        title='Test'
        selected={true}
      />,
    );

    const { output } = shallow(component.getByType(BottomNavigatorTabComponent));

    expect(output).toMatchSnapshot();
  });

  it('* text/unselected', () => {
    const component: RenderAPI = render(
      <Mock
        title='Test'
        selected={false}
      />,
    );

    const { output } = shallow(component.getByType(BottomNavigatorTabComponent));

    expect(output).toMatchSnapshot();
  });

  it('* text (styled)', () => {
    const component: RenderAPI = render(
      <Mock
        title='Test'
        titleStyle={{ fontSize: 22, color: 'yellow' }}
      />,
    );

    const { output } = shallow(component.getByType(BottomNavigatorTabComponent));

    expect(output).toMatchSnapshot();
  });

});

describe('@bottom-navigation: component checks', () => {

  const Mock = (props?: BottomTabNavigatorProps): React.ReactElement<ApplicationProviderProps> => {
    return (
      <ApplicationProvider
        mapping={mapping}
        theme={theme}>
        <BottomNavigation {...props} />
      </ApplicationProvider>
    );
  };

  const ChildMock = (props: BottomNavigatorTabProps): React.ReactElement<BottomNavigatorTabProps> => {
    return (
      <BottomNavigationTab {...props} />
    );
  };

  const tabTestId: string = '@tab/last';

  it('* empty', () => {
    const component: RenderAPI = render(
      <Mock children={[]}/>,
    );

    const { output } = shallow(component.getByType(BottomTabNavigatorComponent));

    expect(output).toMatchSnapshot();
  });

  it('* with routes', () => {
    const indicatorStyle: StyleType = { backgroundColor: 'red' };
    const component: RenderAPI = render(
      <Mock indicatorStyle={indicatorStyle}>
        <ChildMock
          title='Screen 1'
          selected={false}
        />
        <ChildMock
          title='Screen 2'
          selected={true}
        />
        <ChildMock
          title='Screen 3'
          selected={false}
        />
      </Mock>,
    );

    const { output } = shallow(component.getByType(BottomTabNavigatorComponent));

    expect(output).toMatchSnapshot();
  });

  it('* current index', () => {
    const component: RenderAPI = render(
      <Mock selectedIndex={1}>
        <ChildMock
          title='Screen 1'
          selected={false}
        />
        <ChildMock
          title='Screen 2'
          selected={true}
        />
        <ChildMock
          title='Screen 3'
          selected={false}
        />
      </Mock>,
    );

    const componentInstance: ReactTestInstance = component.getByType(BottomTabNavigatorComponent);

    expect(componentInstance.props.selectedIndex).toBe(1);
  });

  it('* tab choose', () => {
    const onSelect = jest.fn();

    const component: RenderAPI = render(
      <Mock
        selectedIndex={1}
        onSelect={onSelect}>
        <ChildMock
          title='Screen 1'
          selected={false}
        />
        <ChildMock
          title='Screen 2'
          selected={true}
        />
        <ChildMock
          testID={tabTestId}
          title='Screen 3'
          selected={false}
        />
      </Mock>,
    );

    fireEvent(component.getByTestId(tabTestId), 'select');

    expect(onSelect).toHaveBeenCalled();
  });

  it('* choose selected tab', () => {
    const onSelect = jest.fn();

    const component: RenderAPI = render(
      <Mock
        selectedIndex={1}
        onSelect={onSelect}>
        <ChildMock
          title='Screen 1'
          selected={false}
        />
        <ChildMock
          testID={tabTestId}
          title='Screen 2'
          selected={true}
        />
        <ChildMock
          title='Screen 3'
          selected={false}
        />
      </Mock>,
    );

    fireEvent(component.getByTestId(tabTestId), 'select');

    expect(onSelect).not.toHaveBeenCalled();
  });

  it('* additional appearance', () => {
    const component: RenderAPI = render(
      <Mock appearance='noIndicator'>
        <ChildMock
          title='Screen 1'
          selected={false}
        />
        <ChildMock
          title='Screen 2'
          selected={true}
        />
        <ChildMock
          testID={tabTestId}
          title='Screen 3'
          selected={false}
        />
      </Mock>,
    );

    const { output } = shallow(component.getByType(BottomTabNavigatorComponent));

    expect(output).toMatchSnapshot();
  });

});

