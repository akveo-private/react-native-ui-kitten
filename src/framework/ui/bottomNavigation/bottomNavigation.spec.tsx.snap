// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@bottom-navigation: component checks * additional appearance 1`] = `
<View
  appearance="noIndicator"
  dispatch={[Function]}
  selectedIndex={0}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderTopColor": "#A6AEBD",
        "borderTopWidth": 1,
        "paddingVertical": 16,
      },
      undefined,
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 750,
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
>
  <ChildMock
    onSelect={[Function]}
    selected={true}
    style={
      Object {
        "flex": 1,
      }
    }
    title="Screen 1"
  />
  <ChildMock
    onSelect={[Function]}
    selected={false}
    style={
      Object {
        "flex": 1,
      }
    }
    title="Screen 2"
  />
  <ChildMock
    onSelect={[Function]}
    selected={false}
    style={
      Object {
        "flex": 1,
      }
    }
    testID="@tab/last"
    title="Screen 3"
  />
</View>
`;

exports[`@bottom-navigation: component checks * empty 1`] = `
<View
  appearance="default"
  dispatch={[Function]}
  selectedIndex={0}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderTopColor": "#A6AEBD",
        "borderTopWidth": 1,
        "paddingVertical": 16,
      },
      undefined,
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 750,
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
>
  <TabBarIndicator
    animationDuration={200}
    positions={0}
    selectedPosition={0}
    style={
      Array [
        Object {
          "backgroundColor": "#3366FF",
          "height": 3,
        },
        Object {
          "position": "absolute",
        },
      ]
    }
  />
</View>
`;

exports[`@bottom-navigation: component checks * with routes 1`] = `
<View
  appearance="default"
  dispatch={[Function]}
  selectedIndex={0}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderTopColor": "#A6AEBD",
        "borderTopWidth": 1,
        "paddingVertical": 16,
      },
      undefined,
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 750,
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
>
  <TabBarIndicator
    animationDuration={200}
    positions={3}
    selectedPosition={0}
    style={
      Array [
        Object {
          "backgroundColor": "#3366FF",
          "height": 3,
        },
        Object {
          "position": "absolute",
        },
      ]
    }
  />
  <ChildMock
    onSelect={[Function]}
    selected={true}
    style={
      Object {
        "flex": 1,
      }
    }
    title="Screen 1"
  />
  <ChildMock
    onSelect={[Function]}
    selected={false}
    style={
      Object {
        "flex": 1,
      }
    }
    title="Screen 2"
  />
  <ChildMock
    onSelect={[Function]}
    selected={false}
    style={
      Object {
        "flex": 1,
      }
    }
    title="Screen 3"
  />
</View>
`;

exports[`@bottom-navigation-tab: component checks * empty 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  onPress={[Function]}
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
/>
`;

exports[`@bottom-navigation-tab: component checks * text/selected 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  onPress={[Function]}
  selected={true}
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Test"
>
  <Text
    style={
      Object {
        "color": "#3366FF",
        "fontWeight": "600",
      }
    }
  >
    Test
  </Text>
</TouchableOpacity>
`;

exports[`@bottom-navigation-tab: component checks * text/unselected 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  onPress={[Function]}
  selected={false}
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
  title="Test"
>
  <Text
    style={
      Object {
        "color": "#8992A3",
        "fontWeight": "600",
      }
    }
  >
    Test
  </Text>
</TouchableOpacity>
`;

exports[`@bottom-navigation-tab: component checks * with icon 1`] = `
<TouchableOpacity
  activeOpacity={1}
  appearance="default"
  dispatch={[Function]}
  icon={[Function]}
  onPress={[Function]}
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
  theme={
    Object {
      "blue-dark": "#2541CC",
      "blue-pitch-dark": "#0D1C2E",
      "blue-primary": "#3366FF",
      "gray-100": "#f7f8fa",
      "gray-200": "#edf0f5",
      "gray-300": "#c8cedb",
      "gray-400": "#a6aebd",
      "gray-dark": "#8992A3",
      "gray-highlight": "#EDF0F5",
      "gray-light": "#DDE1EB",
      "gray-primary": "#A6AEBD",
      "pink-primary": "#FF3D71",
      "text-primary": "#000000",
      "text-primary-inverse": "#ffffff",
    }
  }
>
  <Component
    source={
      Object {
        "uri": "https://akveo.github.io/eva-icons/fill/png/128/star.png",
      }
    }
    style={
      Object {
        "height": 40,
        "marginBottom": undefined,
        "tintColor": "#8992A3",
        "width": 40,
      }
    }
  />
</TouchableOpacity>
`;
